Python/bytecodes.c

```diff
@@ -285,11 +285,25 @@ dummy_func(
         }
 
         family(LOAD_CONST, 0) = {
+            LOAD_CONST_MORTAL,
             LOAD_CONST_IMMORTAL,
         };
 
-        pure inst(LOAD_CONST, (-- value)) {
-            value = PyStackRef_FromPyObjectNew(GETITEM(FRAME_CO_CONSTS, oparg));
+        inst(LOAD_CONST, (-- value)) {
+            /* We can't do this in the bytecode compiler as
+             * marshalling can intern strings and make them immortal. */
+            PyObject *obj = GETITEM(FRAME_CO_CONSTS, oparg);
+            value = PyStackRef_FromPyObjectNew(obj);
+#if ENABLE_SPECIALIZATION
+            if (this_instr->op.code == LOAD_CONST) {
+                this_instr->op.code = _Py_IsImmortal(obj) ? LOAD_CONST_IMMORTAL : LOAD_CONST_MORTAL;
+            }
+#endif
+        }
```

Comment:
Why the check for `this_instr->op.code == LOAD_CONST`?

Also, minor, but maybe a bit more idiomatic with the rest of the bytecode definitions to split the specializing part out and reuse the code for loading the const? I think this would also make `LOAD_CONST` a viable tier two instruction (which it currently isn't). Though actually tracing through one should be rare, I can think of cases where it might happen.

```suggestion
        specializing op(_SPECIALIZE_LOAD_CONST, (--)) {
            /* We can't do this in the bytecode compiler as
             * marshalling can intern strings and make them immortal. */
#if ENABLE_SPECIALIZATION
            this_instr->op.code = _Py_IsImmortal(obj) ? LOAD_CONST_IMMORTAL : LOAD_CONST_MORTAL;
#endif
        }
        
        macro(LOAD_CONST) = _SPECIALIZE_LOAD_CONST + LOAD_CONST_MORTAL;
```

Reply:
> Why the check for this_instr->op.code == LOAD_CONST?

Instrumentation. The opcode might be `INSTRUMENTED_LINE` for example.

Reply:
> split the specializing part out

The specializing part needs to load the value to test it, so separating out the specialization just duplicates that.